spec/lib/bulk_data/cache_spec.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+RSpec.describe BulkData::Cache do
+  describe ".write" do
+    it "adds an entry that expires in 24 hours" do
+      BulkData::Cache.write("key", "value")
+      expect(BulkData::Cache.cache.fetch("key")).to eq("value")
+
+      travel_to(25.hours.from_now) do
+        expect(BulkData::Cache.cache.fetch("key")).to be_nil
+      end
+    end
+
+    it "sets the current time as the age, which also expires" do
+      freeze_time do
+        BulkData::Cache.write("key", "value")
+        expect(BulkData::Cache.cache.fetch("key:created")).to eq(Time.current)
+      end
+
+      travel_to(25.hours.from_now) do
+        expect(BulkData::Cache.cache.fetch("key:created")).to be_nil
+      end
+    end
+  end
+
+  describe ".fetch" do
+    it "returns a value when one exists" do
+      BulkData::Cache.write("key", "value")
+      expect(BulkData::Cache.fetch("key")).to eq("value")
+    end
+
+    it "raises a NoEntryError when there isn't an entry" do
+      expect { BulkData::Cache.fetch("key") }
+        .to raise_error(BulkData::Cache::NoEntryError)
+    end
+  end
+
+  describe ".wrote_at" do
+    it "returns the time the entry was created" do
+      time = 2.hours.ago.change(usec: 0)
+      travel_to(time) { BulkData::Cache.write("key", "value") }
+      expect(BulkData::Cache.wrote_at("key")).to eq(time)
+    end
+
+    it "returns nil when the entry doesn't exist" do
+      expect(BulkData::Cache.wrote_at("key")).to be_nil
+    end
+  end
+
+  describe ".wrote_after?" do
+    it "returns true if the cache was wrote after the time" do
+      travel_to(3.minutes.ago) { BulkData::Cache.write("key", "value") }
+      expect(BulkData::Cache.wrote_after?("key", 5.minutes.ago)).to be true
+    end
+
+    it "returns false if the cache was wrote before the time" do
+      travel_to(10.minutes.ago) { BulkData::Cache.write("key", "value") }
+      expect(BulkData::Cache.wrote_after?("key", 5.minutes.ago)).to be false
+    end
+
+    it "returns false if a time isn't provided" do
+      BulkData::Cache.write("key", "value")
+      expect(BulkData::Cache.wrote_after?("key", nil)).to be false
```

Comment:
Yup that's how it's called in GovernmentRepository.

Reply:
If that's the case, can we call it without an explicit `nil`? Think it's more self-explanatory to have the method definition as `def wrote_after?(key, time: nil)` than `def wrote_after?(key, time)` as the former makes it clear that `time` is optional.

```suggestion
      expect(BulkData::Cache.wrote_after?("key")).to be false
```

I wonder if it's better to have two distinct methods rather than having forks of logic inside this one. Think it would remove any ambiguity and also make it read better, as `wrote_after?("key")` doesn't really make sense if no time is provided.

```ruby
    def self.written?(key)
      wrote_at(key).present?
    end

    def self.wrote_after?(key, time)
      created = wrote_at(key)
      return false unless created

      created > time
    end
```

Reply:
I think it's better to look at this from the execution point:

https://github.com/alphagov/content-publisher/blob/b89f071881832704a41b67cc796e8f80d585f4dd/lib/bulk_data/government_repository.rb#L33-L34

What I'm trying to avoid is a need to have to but guard clauses around any call to the method

so with your suggestion it'd be `return if older_than && Cache.wrote_after?(CACHE_KEY, older_than)` which just feels a bit more than necessary - what do you think? can switch it to that if you think it's an improvement just felt it was a bit of boilerplate on the caller side.

Reply:
Ah, I see why you've done it, looking at it from that perspective. Personally, I think it would be an improvement to make this line slightly more verbose as you've suggested, if it then means we have simpler code in the `wrote_after?` method (and one less test) and less variability to keep in your head while reading it.

I don't feel too strongly about it though, and appreciate it would take a bit of work to undo and you're on a tight deadline.

Reply:
I've gone with your preferred approach.